datalad/support/tests/test_annexrepo.py

```diff
@@ -2495,4 +2496,63 @@ def check_commit_annex_commit_changed(unlock, path):
 
 def test_commit_annex_commit_changed():
     for unlock in True, False:
-        yield check_commit_annex_commit_changed, unlock
\ No newline at end of file
+        yield check_commit_annex_commit_changed, unlock
+
+
+@with_tempfile(mkdir=True)
+def check_files_split_exc(cls, topdir):
+    from glob import glob
+    r = cls(topdir)
+    # absent files -- should not crash with "too long" but some other more
+    # meaningful exception
+    files = ["f" * 100 + "%04d" % f for f in range(100000)]
+    if isinstance(r, AnnexRepo):
+        # Annex'es add first checks for what is being added and does not fail
+        # for non existing files either ATM :-/  TODO: make consistent etc
+        r.add(files)
+    else:
+        with assert_raises(Exception) as ecm:
+            r.add(files)
+        assert_not_in('too long', str(ecm.exception))
+        assert_not_in('too many', str(ecm.exception))
+
+
+def test_files_split_exc():
+    for cls in GitRepo, AnnexRepo:
+        yield check_files_split_exc, cls
+
+
+_HEAVY_TREE = {
+    # might already run into 'filename too long' on windows probably
```

Comment:
Just to check my understanding/quick math: `test_files_split` is intended to be a smoke test that, given the system/configuration, may or may not have failed _before_ this PR?  (When I execute this test on top of master it passes on my system where SC_ARG_MAX matches our CMD_MAX_ARG_HARDCODED for GNU/Linux.)

Reply:
nope -- fails for me also with TMPDIR=/tmp
<details>
<summary>my case</summary>

```shell
(git)hopa:~datalad/datalad-master[enh-tst-files-split]git
$> DATALAD_TESTS_SSH=1 TMPDIR=/tmp python -m nose -v ./datalad/support/tests/test_annexrepo.py:test_files_split datalad.support.tests.test_annexrepo.test_files_split(<class 'datalad.support.gitrepo.GitRepo'>,) ... ERROR
datalad.support.tests.test_annexrepo.test_files_split(<class 'datalad.support.annexrepo.AnnexRepo'>,) ... ERROR

======================================================================
ERROR: datalad.support.tests.test_annexrepo.test_files_split(<class 'datalad.support.gitrepo.GitRepo'>,)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/yoh/proj/datalad/datalad-master/datalad/tests/utils.py", line 442, in newfunc
    return t(*(arg + (d,)), **kw)
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/tests/test_annexrepo.py", line 2536, in check_files_split
    r.add(files)
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/gitrepo.py", line 313, in newfunc
    result = func(self, files_new, *args, **kwargs)
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/gitrepo.py", line 1057, in add
    update=update))
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/gitrepo.py", line 1085, in add_
    to_options(update=update) + ['--verbose']
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/gitrepo.py", line 313, in newfunc
    result = func(self, files_new, *args, **kwargs)
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/gitrepo.py", line 1747, in _git_custom_command
    expect_fail=expect_fail)
  File "/home/yoh/proj/datalad/datalad-master/datalad/cmd.py", line 682, in run
    cmd, env=self.get_git_environ_adjusted(env), *args, **kwargs)
  File "/home/yoh/proj/datalad/datalad-master/datalad/cmd.py", line 496, in run
    stdin=stdin)
  File "/usr/lib/python2.7/subprocess.py", line 394, in __init__
    errread, errwrite)
  File "/usr/lib/python2.7/subprocess.py", line 1047, in _execute_child
    raise child_exception
OSError: [Errno 7] Argument list too long
-------------------- >> begin captured logging << --------------------
datalad.utils: Level 5: Importing datalad.utils
datalad.utils: DEBUG: Maximal length of cmdline string (adjusted for safety margin): 1048576
datalad.utils: Level 5: Done importing datalad.utils
datalad.cmd: Level 9: Will use git under '/usr/lib/git-annex.linux' (no adjustments to PATH if empty string)
datalad.cmd: Level 9: Running: ['git', 'version']
datalad.cmd: Level 8: Finished running ['git', 'version'] with status 0
datalad.cmd: Level 9: Running: ['git', 'config', '-z', '-l', '--show-origin']
datalad.cmd: Level 8: Finished running ['git', 'config', '-z', '-l', '--show-origin'] with status 0
datalad.ui: Level 5: Starting importing ui
datalad.ui.dialog: Level 5: Starting importing ui.dialog
datalad.ui.dialog: Level 5: Done importing ui.dialog
datalad.ui: Level 5: Initiating UI switcher
datalad.ui: DEBUG: UI set to ConsoleLog(out=<StreamWrapper>)
datalad.ui: Level 5: Done importing ui
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: datalad.support.tests.test_annexrepo.test_files_split(<class 'datalad.support.annexrepo.AnnexRepo'>,)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/yoh/proj/datalad/datalad-master/datalad/tests/utils.py", line 442, in newfunc
    return t(*(arg + (d,)), **kw)
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/tests/test_annexrepo.py", line 2536, in check_files_split
    r.add(files)
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/gitrepo.py", line 313, in newfunc
    result = func(self, files_new, *args, **kwargs)
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/annexrepo.py", line 1470, in add
    git_options=git_options, annex_options=annex_options, update=update
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/annexrepo.py", line 1545, in add_
    for rec in to_be_added_recs
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/annexrepo.py", line 1544, in <dictcomp>
    rec['file']: self.get_file_size(rec['file'])
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/annexrepo.py", line 1515, in _get_to_be_added_recs
    files, git_options=_git_options, update=update):
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/gitrepo.py", line 1085, in add_
    to_options(update=update) + ['--verbose']
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/annexrepo.py", line 1256, in _git_custom_command
    return super(AnnexRepo, self)._git_custom_command(*args, **kwargs)
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/gitrepo.py", line 313, in newfunc
    result = func(self, files_new, *args, **kwargs)
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/gitrepo.py", line 1747, in _git_custom_command
    expect_fail=expect_fail)
  File "/home/yoh/proj/datalad/datalad-master/datalad/cmd.py", line 682, in run
    cmd, env=self.get_git_environ_adjusted(env), *args, **kwargs)
  File "/home/yoh/proj/datalad/datalad-master/datalad/cmd.py", line 496, in run
    stdin=stdin)
  File "/usr/lib/python2.7/subprocess.py", line 394, in __init__
    errread, errwrite)
  File "/usr/lib/python2.7/subprocess.py", line 1047, in _execute_child
    raise child_exception
OSError: [Errno 7] Argument list too long

----------------------------------------------------------------------
Ran 2 tests in 3.093s

FAILED (errors=2)
```
</details>

I have pushed branch enh-tst-files-split to my fork which is on top of the master has only those tests added before I cherry picked it into this bug fix branch.
Could you check to what number of files or dirs you need to increase to cause the failure?